File: web/src/domain/huaweiHealth.ts
/**
 * Lectura de la exportacion de Huawei Health.
 *
 * Huawei exporta un JSON por entrenamiento grabado (carpeta `Motion path detail data`), con el tipo
 * de deporte, el inicio, el fin, la duracion, las calorias, la distancia y la frecuencia cardiaca.
 * No exporta series con peso y reps: esos entrenamientos entran como sesiones sin series.
 *
 * El lector es tolerante a proposito: la exportacion trae tambien archivos de sueno, pasos y
 * estres, y el formato varia entre versiones de la app. Todo lo que no se reconoce se ignora.
 */

import {
  HUAWEI_SOURCE,
  importedWorkoutNote,
  workoutKey,
  type ImportedWorkout,
} from '@/domain/importedWorkout';

/** Marca de origen en la nota de una sesión importada de Huawei Health. */
export const HUAWEI_NOTE_PREFIX = HUAWEI_SOURCE;

/** `true` si la sesión vino de una importación. */
export { isImportedNote as isHuaweiNote } from '@/domain/importedWorkout';

/** Nota de la sesión importada, con el origen y los datos disponibles. */
export const huaweiNote = importedWorkoutNote;

export interface HuaweiParseResult {
  readonly workouts: readonly ImportedWorkout[];
  readonly filesRead: number;
  readonly filesSkipped: number;
}

/** Nombres de los tipos de deporte que usa la exportacion. */
const SPORT_NAMES: Record<number, string> = {
  2: 'Senderismo',
  3: 'Bicicleta',
  4: 'Running',
  5: 'Caminata',
  102: 'Natación en pileta',
  103: 'Bicicleta fija',
  104: 'Natación en aguas abiertas',
  112: 'Remo',
  117: 'Entrenamiento',
  118: 'Running',
  147: 'Entrenamiento de fuerza',
  281: 'Caminata bajo techo',
  282: 'Senderismo',
};

/** Escala de la exportacion: las calorias vienen multiplicadas por mil. */
const CALORIE_SCALE = 1000;
/** Epoch en segundos del siglo XXI: por debajo se asume segundos, no milisegundos. */
const MAX_SECONDS_EPOCH = 4_102_444_800;
/** Segundos de un dia: separa una duracion en segundos de una en milisegundos. */
const SECONDS_PER_DAY = 86_400;

/** Lee varios archivos y devuelve los entrenamientos reconocidos, sin repetidos. */
export function parseHuaweiExport(contents: readonly string[]): HuaweiParseResult {
  const byKey = new Map<string, ImportedWorkout>();
  let filesRead = 0;
  let filesSkipped = 0;

  for (const content of contents) {
    let workouts: ImportedWorkout[];
    try {
      workouts = parseHuaweiFile(content);
    } catch {
      filesSkipped += 1;
      continue;
    }
    filesRead += 1;
    for (const workout of workouts) {
      // La exportacion triplica los registros: el primero que aparece gana.
      const key = workoutKey(workout);
      if (!byKey.has(key)) byKey.set(key, workout);
    }
  }

  return {
    workouts: [...byKey.values()].sort((a, b) => b.startedAtMs - a.startedAtMs),
    filesRead,
    filesSkipped,
  };
}

/**
 * Lee un archivo de la exportación.
 *
 * Primero intenta el camino normal (lista u objeto). Si no encuentra entrenamientos, recorre el
 * texto buscando objetos balanceados: así entran los archivos con varios objetos concatenados (la
 * exportación los parte con marcadores de resincronización) y los que traen una comilla suelta
 * dentro del blob de sensores, que rompe el JSON.
 */
export function parseHuaweiFile(content: string): ImportedWorkout[] {
  const trimmed = content.trim();
  if (trimmed === '') return [];

  // Camino rápido: el archivo entero es JSON válido (una lista o un objeto).
  const direct: ImportedWorkout[] = [];
  try {
    collect(JSON.parse(trimmed), direct);
  } catch {
    // Un archivo con varios objetos seguidos no es JSON válido entero: se recorre abajo.
  }

  const candidates = balancedObjects(trimmed);
  if (direct.length > 0 && candidates.length <= 1) return direct;

  // Varios objetos seguidos (o JSON roto por una comilla suelta): se leen uno por uno.
  const scanned: ImportedWorkout[] = [];
  for (const candidate of candidates) {
    let parsed: unknown;
    try {
      parsed = JSON.parse(candidate);
    } catch {
      try {
        parsed = JSON.parse(repairAttributeQuotes(candidate));
      } catch {
        continue;
      }
    }
    collect(parsed, scanned);
  }

  const result = scanned.length > 0 ? scanned : direct;
  const unique = new Map<string, ImportedWorkout>();
  for (const workout of result) unique.set(workoutKey(workout), workout);
  return [...unique.values()];
}

/**
 * Objetos JSON balanceados del texto, respetando los literales de texto.
 *
 * Cubre archivos con varios objetos seguidos, con o sin saltos de línea, y listas: en todos los
 * casos los objetos se recortan por profundidad de llaves.
 */
export function balancedObjects(text: string): string[] {
  const objects: string[] = [];
  let depth = 0;
  let start = -1;
  let inString = false;
  let escaped = false;

  for (let index = 0; index < text.length; index += 1) {
    const char = text[index];
    if (inString) {
      if (escaped) escaped = false;
      else if (char === '\\') escaped = true;
      else if (char === '"') inString = false;
      continue;
    }
    if (char === '"') {
      inString = true;
      continue;
    }
    if (char === '{') {
      if (depth === 0) start = index;
      depth += 1;
      continue;
    }
    if (char === '}') {
      depth -= 1;
      if (depth <= 0 && start >= 0) {
        objects.push(text.slice(start, index + 1));
        start = -1;
        depth = 0;
      }
    }
  }
  return objects;
}

/**
 * Quita las comillas sueltas dentro del campo `attribute`.
 *
 * La exportación guarda ahí la telemetría como texto y a veces aparece una comilla sin escapar, que
 * invalida el archivo entero. El valor real termina en la comilla seguida de coma o cierre.
 */
export function repairAttributeQuotes(text: string): string {
  if (!ATTRIBUTE_FIELD.test(text)) return text;
  ATTRIBUTE_FIELD.lastIndex = 0;

  let result = '';
  let index = 0;
  while (index < text.length) {
    ATTRIBUTE_FIELD.lastIndex = index;
    const match = ATTRIBUTE_FIELD.exec(text);
    if (match === null) break;

    const valueStart = match.index + match[0].length;
    result += text.slice(index, valueStart);

    let cursor = valueStart;
    let end = -1;
    while (cursor < text.length) {
      if (text[cursor] === '"') {
        let probe = cursor + 1;
        while (probe < text.length && /\s/.test(text[probe] ?? '')) probe += 1;
        const next = text[probe];
        if (next === undefined || next === ',' || next === '}' || next === ']') {
          end = cursor;
          break;
        }
      }
      cursor += 1;
    }
    if (end < 0) {
      index = valueStart;
      continue;
    }
    result += text.slice(valueStart, end).replaceAll('"', '');
    index = end;
  }
  result += text.slice(index);
  return result;
}

const ATTRIBUTE_FIELD = /"attribute"\s*:\s*"/gi;

function collect(node: unknown, out: ImportedWorkout[]): void {
  if (Array.isArray(node)) {
    for (const item of node) collect(item, out);
    return;
  }
  if (node !== null && typeof node === 'object') {
    const record = node as Record<string, unknown>;
    const workout = toWorkout(record);
    if (workout !== null) out.push(workout);
    for (const value of Object.values(record)) {
      if (value !== null && typeof value === 'object') collect(value, out);
    }
  }
}

/**
 * Convierte un objeto en entrenamiento, o `null` si no lo parece.
 *
 * No alcanza con tener fecha: la exportacion tambien trae sueno, pasos y estres con sus propias
 * marcas de tiempo. Se pide ademas alguna senal de deporte (tipo, nombre, distancia, calorias o
 * frecuencia cardiaca), que es lo que distingue un entrenamiento del resto.
 */
function toWorkout(node: Record<string, unknown>): ImportedWorkout | null {
  const startedAt = readTimestamp(node, ['startTime', 'start_time', 'beginTime', 'start']);
  if (startedAt === null || !looksLikeWorkout(node)) return null;

  const durationMs = readDuration(node);
  const explicitEnd = readTimestamp(node, ['endTime', 'end_time', 'finishTime', 'end']);
  // Un entrenamiento dura: los archivos por minuto traen fecha y calorías, pero no duración.
  if (durationMs === null && explicitEnd === null) return null;
  const finishedAt = explicitEnd ?? (durationMs === null ? null : startedAt + durationMs);

  const attribute = readString(node, ['attribute', 'attributes']);
  const summaryHeartRate = readNumber(node, ['avgHeartRate', 'averageHeartRate', 'avg_heart_rate']);
  const track = summaryHeartRate === null ? trackHeartRate(attribute) : null;

  const sportType = readNumber(node, ['sportType', 'sport_type', 'activityType', 'exerciseType']);
  const explicitName = readString(node, ['sportName', 'activityName', 'name']);
  const sportName =
    explicitName ?? (sportType === null ? null : SPORT_NAMES[sportType]) ?? 'Entrenamiento';
  const calories = readNumber(node, ['totalCalories', 'calories', 'total_calories']);

  return {
    source: HUAWEI_SOURCE,
    recordId: readString(node, ['recordId', 'record_id', 'id']),
    startedAtMs: startedAt,
    finishedAtMs: finishedAt,
    sportType: sportType === null ? null : Math.round(sportType),
    sportName,
    durationMs: durationMs ?? (finishedAt === null ? null : Math.max(0, finishedAt - startedAt)),
    distanceM: readNumber(node, ['totalDistance', 'distance', 'total_distance']),
    calories: calories === null ? null : calories / CALORIE_SCALE,
    steps: readNumber(node, ['totalSteps', 'steps']),
    averageHeartRate: summaryHeartRate ?? track?.average ?? null,
    maxHeartRate: readNumber(node, ['maxHeartRate', 'max_heart_rate']) ?? track?.max ?? null,
    elevationGainM: null,
  };
}

/**
 * Frecuencia cardiaca derivada del blob de sensores (`attribute`).
 *
 * La exportacion documenta `avgHeartRate` y `maxHeartRate`, pero en la practica vienen vacios: el
 * dato real esta en los segmentos `tp=h-r;k=<minuto>;v=<pulsaciones>;` del blob.
 */
export function trackHeartRate(
  attribute: string | null
): { average: number; max: number } | null {
  if (attribute === null || attribute.trim() === '') return null;

  const readings: number[] = [];
  const lower = attribute.toLowerCase();
  const tag = 'tp=h-r;';
  let index = 0;

  while (index < attribute.length) {
    const segmentStart = lower.indexOf(tag, index);
    if (segmentStart < 0) break;
    const contentStart = segmentStart + tag.length;
    const nextSegment = lower.indexOf('tp=', contentStart);
    const end = nextSegment < 0 ? attribute.length : nextSegment;

    for (const match of attribute.slice(contentStart, end).matchAll(HEART_RATE_PAIR)) {
      const value = Number(match[2]);
      if (Number.isFinite(value)) readings.push(value);
    }
    index = end;
  }

  if (readings.length === 0) return null;
  const total = readings.reduce((sum, value) => sum + value, 0);
  return { average: total / readings.length, max: Math.max(...readings) };
}

/** Lectura de frecuencia cardiaca dentro de un segmento `h-r`: `k=<minuto>;v=<pulsaciones>;`. */
const HEART_RATE_PAIR = /k=([-\d.]+);v=([-\d.]+);/gi;

/** Senales de que el objeto es un entrenamiento y no otra serie de datos de la exportacion. */
function looksLikeWorkout(node: Record<string, unknown>): boolean {
  if (readNumber(node, ['sportType', 'sport_type', 'activityType', 'exerciseType']) !== null) {
    return true;
  }
  if (readString(node, ['sportName', 'activityName']) !== null) return true;
  if (readNumber(node, ['totalDistance', 'distance', 'total_distance']) !== null) return true;
  if (readNumber(node, ['totalCalories', 'calories', 'total_calories']) !== null) return true;
  return (
    readNumber(node, ['avgHeartRate', 'averageHeartRate', 'maxHeartRate', 'max_heart_rate']) !== null
  );
}

/**
 * Lee una fecha: epoch en milisegundos o segundos, o texto ISO / `yyyy-MM-dd HH:mm:ss`.
 *
 * La exportacion usa milisegundos, pero algunas versiones exportan segundos; se distingue por
 * magnitud.
 */
function readTimestamp(node: Record<string, unknown>, keys: readonly string[]): number | null {
  for (const key of keys) {
    if (!(key in node)) continue;
    const raw = node[key];
    if (typeof raw === 'number' && Number.isFinite(raw)) return normalizeEpoch(raw);
    if (typeof raw === 'string') {
      const text = raw.trim();
      if (text === '') continue;
      if (/^\d+$/.test(text)) return normalizeEpoch(Number(text));
      // Sin zona explicita la fecha se interpreta como UTC, igual que en Android: si no, el
      // resultado depende de la zona del dispositivo.
      const iso = text.includes('T') ? text : text.replace(' ', 'T');
      const hasZone = /([zZ]|[+-]\d{2}:?\d{2})$/.test(iso);
      const parsed = Date.parse(hasZone ? iso : `${iso}Z`);
      if (!Number.isNaN(parsed)) return parsed;
    }
  }
  return null;
}

function normalizeEpoch(value: number): number {
  const rounded = Math.trunc(value);
  return rounded >= 1 && rounded <= MAX_SECONDS_EPOCH ? rounded * 1000 : rounded;
}

/**
 * Duracion en milisegundos.
 *
 * `totalTime` ya viene en milisegundos (la exportacion divide por mil para obtener segundos); las
 * claves alternativas suelen venir en segundos, asi que se distinguen por magnitud.
 */
function readDuration(node: Record<string, unknown>): number | null {
  const total = readNumber(node, ['totalTime', 'total_time']);
  if (total !== null) return Math.trunc(total);

  const explicitMs = readNumber(node, ['durationMs', 'duration_ms']);
  if (explicitMs !== null) return Math.trunc(explicitMs);

  const seconds = readNumber(node, ['duration', 'exerciseTime']);
  if (seconds !== null) {
    return seconds < SECONDS_PER_DAY ? Math.trunc(seconds * 1000) : Math.trunc(seconds);
  }
  return null;
}

function readString(node: Record<string, unknown>, keys: readonly string[]): string | null {
  for (const key of keys) {
    const value = node[key];
    if (typeof value === 'string' && value.trim() !== '') return value.trim();
    if (typeof value === 'number' && Number.isFinite(value)) return String(value);
  }
  return null;
}

function readNumber(node: Record<string, unknown>, keys: readonly string[]): number | null {
  for (const key of keys) {
    const value = node[key];
    if (typeof value === 'number' && Number.isFinite(value)) return value;
    if (typeof value === 'string') {
      const parsed = Number(value.trim().replace(',', '.'));
      if (Number.isFinite(parsed)) return parsed;
    }
  }
  return null;
}

File: android/app/src/main/java/com/fitlog/app/domain/HuaweiHealth.kt
package com.fitlog.app.domain

import org.json.JSONArray
import org.json.JSONObject
import org.json.JSONTokener
import java.text.SimpleDateFormat
import java.util.Locale
import java.util.TimeZone

/**
 * Lectura de la exportacion de Huawei Health.
 *
 * Huawei exporta un JSON por entrenamiento grabado (carpeta `Motion path detail data`), con el tipo
 * de deporte, el inicio, el fin, la duracion, las calorias, la distancia y la frecuencia cardiaca.
 * No exporta series con peso y reps: esos entrenamientos entran como sesiones sin series.
 *
 * El lector es tolerante a proposito: la exportacion trae tambien archivos de sueno, pasos y
 * estres, y el formato varia entre versiones de la app. Todo lo que no se reconoce se ignora.
 */
object HuaweiHealth {

    /** Resultado de leer un conjunto de archivos. */
    data class ParseResult(
        val workouts: List<ImportedWorkout>,
        val filesRead: Int,
        val filesSkipped: Int,
    ) {
        val isEmpty: Boolean get() = workouts.isEmpty()
    }

    /** Nombres de los tipos de deporte que usa la exportacion. */
    private val SPORT_NAMES = mapOf(
        2 to "Senderismo",
        3 to "Bicicleta",
        4 to "Running",
        5 to "Caminata",
        102 to "Natación en pileta",
        103 to "Bicicleta fija",
        104 to "Natación en aguas abiertas",
        112 to "Remo",
        117 to "Entrenamiento",
        118 to "Running",
        147 to "Entrenamiento de fuerza",
        281 to "Caminata bajo techo",
        282 to "Senderismo",
    )

    /** Escala de la exportacion: las calorias vienen multiplicadas por mil. */
    private const val CALORIE_SCALE = 1000.0

    /** Lee varios archivos y devuelve los entrenamientos reconocidos, sin repetidos. */
    fun parse(contents: List<String>): ParseResult {
        val byKey = linkedMapOf<String, ImportedWorkout>()
        var read = 0
        var skipped = 0

        for (content in contents) {
            val workouts = runCatching { parseFile(content) }.getOrNull()
            if (workouts == null) {
                skipped += 1
                continue
            }
            read += 1
            for (workout in workouts) {
                // La exportacion triplica los registros: el primero que aparece gana.
                byKey.putIfAbsent(workout.key, workout)
            }
        }

        return ParseResult(
            workouts = byKey.values.sortedByDescending { it.startedAtMs },
            filesRead = read,
            filesSkipped = skipped,
        )
    }

    /**
     * Lee un archivo de la exportacion.
     *
     * Primero intenta el camino normal (lista u objeto). Si no encuentra entrenamientos, recorre el
     * texto buscando objetos balanceados: asi entran los archivos con varios objetos concatenados
     * (la exportacion los parte con marcadores de resincronizacion) y los que traen una comilla
     * suelta dentro del blob de sensores, que rompe el JSON.
     */
    fun parseFile(content: String): List<ImportedWorkout> {
        val trimmed = content.trim()
        if (trimmed.isEmpty()) return emptyList()

        // Camino rapido: el archivo entero es JSON valido (una lista o un objeto).
        val root = runCatching { JSONTokener(trimmed).nextValue() }.getOrNull()
        val direct = mutableListOf<ImportedWorkout>()
        when (root) {
            is JSONArray -> collect(root, direct)
            is JSONObject -> collect(root, direct)
        }

        val candidates = balancedObjects(trimmed)
        if (direct.isNotEmpty() && candidates.size <= 1) return direct

        // Varios objetos seguidos (o JSON roto por una comilla suelta): se leen uno por uno.
        val scanned = mutableListOf<ImportedWorkout>()
        for (candidate in candidates) {
            val parsed = runCatching { JSONObject(candidate) }.getOrNull()
                ?: runCatching { JSONObject(repairAttributeQuotes(candidate)) }.getOrNull()
                ?: continue
            collect(parsed, scanned)
        }

        val result = if (scanned.isNotEmpty()) scanned else direct
        return result.distinctBy { it.key }
    }

    /**
     * Objetos JSON balanceados del texto, respetando los literales de texto.
     *
     * Cubre archivos con varios objetos seguidos, con o sin saltos de linea, y listas: en todos los
     * casos los objetos se recortan por profundidad de llaves.
     */
    private fun balancedObjects(text: String): List<String> {
        val objects = mutableListOf<String>()
        var depth = 0
        var start = -1
        var inString = false
        var escaped = false

        for (index in text.indices) {
            val char = text[index]
            if (inString) {
                when {
                    escaped -> escaped = false
                    char == '\\' -> escaped = true
                    char == '"' -> inString = false
                }
                continue
            }
            when (char) {
                '"' -> inString = true
                '{' -> {
                    if (depth == 0) start = index
                    depth += 1
                }
                '}' -> {
                    depth -= 1
                    if (depth <= 0 && start >= 0) {
                        objects += text.substring(start, index + 1)
                        start = -1
                        depth = 0
                    }
                }
            }
        }
        return objects
    }

    /**
     * Quita las comillas sueltas dentro del campo `attribute`.
     *
     * La exportacion guarda ahi la telemetria como texto y a veces aparece una comilla sin escapar,
     * que invalida el archivo entero. El valor real termina en la comilla seguida de coma o cierre.
     */
    fun repairAttributeQuotes(text: String): String {
        if (!text.contains(ATTRIBUTE_FIELD)) return text

        val builder = StringBuilder()
        var index = 0
        while (index < text.length) {
            val match = ATTRIBUTE_FIELD.find(text, index) ?: break
            val valueStart = match.range.last + 1
            builder.append(text, index, valueStart)

            var cursor = valueStart
            var end = -1
            while (cursor < text.length) {
                if (text[cursor] == '"') {
                    var probe = cursor + 1
                    while (probe < text.length && text[probe].isWhitespace()) probe += 1
                    val closes = probe >= text.length ||
                        text[probe] == ',' || text[probe] == '}' || text[probe] == ']'
                    if (closes) {
                        end = cursor
                        break
                    }
                }
                cursor += 1
            }
            if (end < 0) {
                index = valueStart
                continue
            }
            builder.append(text.substring(valueStart, end).replace("\"", ""))
            index = end
        }
        builder.append(text, index, text.length)
        return builder.toString()
    }

    private val ATTRIBUTE_FIELD = Regex("\"attribute\"\\s*:\\s*\"", RegexOption.IGNORE_CASE)

    private fun collect(node: Any?, out: MutableList<ImportedWorkout>) {
        when (node) {
            is JSONArray -> for (index in 0 until node.length()) collect(node.opt(index), out)
            is JSONObject -> {
                toWorkout(node)?.let { out += it }
                node.keys().forEach { key ->
                    val value = node.opt(key)
                    if (value is JSONArray || value is JSONObject) collect(value, out)
                }
            }
        }
    }

    /**
     * Convierte un objeto en entrenamiento, o `null` si no lo parece.
     *
     * No alcanza con tener fecha: la exportacion tambien trae sueno, pasos y estres con sus propias
     * marcas de tiempo. Se pide ademas alguna senal de deporte (tipo, nombre, distancia, calorias o
     * frecuencia cardiaca), que es lo que distingue un entrenamiento del resto.
     */
    private fun toWorkout(node: JSONObject): ImportedWorkout? {
        val startedAt = readTimestamp(node, "startTime", "start_time", "beginTime", "start")
            ?: return null
        if (!looksLikeWorkout(node)) return null

        val durationMs = readDuration(node)
        val explicitEnd = readTimestamp(node, "endTime", "end_time", "finishTime", "end")
        // Un entrenamiento dura: los archivos por minuto traen fecha y calorias, pero no duracion.
        if (durationMs == null && explicitEnd == null) return null
        val finishedAt = explicitEnd ?: durationMs?.let { startedAt + it }

        val attribute = readString(node, "attribute", "attributes")
        val summaryHeartRate = readDouble(node, "avgHeartRate", "averageHeartRate", "avg_heart_rate")
        val track = if (summaryHeartRate == null) trackHeartRate(attribute) else null

        val sportType = readInt(node, "sportType", "sport_type", "activityType", "exerciseType")
        val sportName = readString(node, "sportName", "activityName", "name")
            ?: sportType?.let { SPORT_NAMES[it] }
            ?: "Entrenamiento"

        return ImportedWorkout(
            recordId = readString(node, "recordId", "record_id", "id"),
            startedAtMs = startedAt,
            finishedAtMs = finishedAt,
            sportType = sportType,
            sportName = sportName,
            durationMs = durationMs ?: finishedAt?.let { (it - startedAt).coerceAtLeast(0L) },
            distanceM = readDouble(node, "totalDistance", "distance", "total_distance"),
            calories = readDouble(node, "totalCalories", "calories", "total_calories")
                ?.let { it / CALORIE_SCALE },
            steps = readInt(node, "totalSteps", "steps"),
            averageHeartRate = summaryHeartRate ?: track?.first,
            maxHeartRate = readDouble(node, "maxHeartRate", "max_heart_rate") ?: track?.second,
        )
    }

    /**
     * Frecuencia cardiaca derivada del blob de sensores (`attribute`).
     *
     * La exportacion documenta `avgHeartRate` y `maxHeartRate`, pero en la practica vienen vacios:
     * el dato real esta en los segmentos `tp=h-r;k=<minuto>;v=<pulsaciones>;` del blob. Devuelve
     * (promedio, maximo), o `null` si no hay lecturas.
     */
    fun trackHeartRate(attribute: String?): Pair<Double, Double>? {
        if (attribute.isNullOrBlank()) return null

        val readings = mutableListOf<Double>()
        var index = 0
        while (index < attribute.length) {
            val segmentStart = attribute.indexOf(HR_TAG, index, ignoreCase = true)
            if (segmentStart < 0) break
            val contentStart = segmentStart + HR_TAG.length
            val nextSegment = attribute.indexOf("tp=", contentStart, ignoreCase = true)
            val end = if (nextSegment < 0) attribute.length else nextSegment
            HEART_RATE_PAIR.findAll(attribute.substring(contentStart, end)).forEach { match ->
                match.groupValues[2].toDoubleOrNull()?.let { readings += it }
            }
            index = end
        }

        if (readings.isEmpty()) return null
        return readings.average() to readings.max()
    }

    private const val HR_TAG = "tp=h-r;"

    /** Lectura de frecuencia cardiaca dentro de un segmento `h-r`: `k=<minuto>;v=<pulsaciones>;`. */
    private val HEART_RATE_PAIR = Regex("k=([-\\d.]+);v=([-\\d.]+);", RegexOption.IGNORE_CASE)

    /** Senales de que el objeto es un entrenamiento y no otra serie de datos de la exportacion. */
    private fun looksLikeWorkout(node: JSONObject): Boolean {
        if (readDouble(node, "sportType", "sport_type", "activityType", "exerciseType") != null) {
            return true
        }
        if (readString(node, "sportName", "activityName") != null) return true
        if (readDouble(node, "totalDistance", "distance", "total_distance") != null) return true
        if (readDouble(node, "totalCalories", "calories", "total_calories") != null) return true
        return readDouble(node, "avgHeartRate", "averageHeartRate", "maxHeartRate") != null
    }

    /**
     * Lee una fecha: epoch en milisegundos o segundos, o texto ISO / `yyyy-MM-dd HH:mm:ss`.
     *
     * La exportacion usa milisegundos, pero algunas versiones exportan segundos; se distingue por
     * magnitud (un epoch en segundos de este siglo tiene diez digitos, uno en milisegundos trece).
     */
    private fun readTimestamp(node: JSONObject, vararg keys: String): Long? {
        for (key in keys) {
            if (!node.has(key)) continue
            val raw = node.opt(key)
            when (raw) {
                is Number -> return normalizeEpoch(raw.toLong())
                is String -> {
                    val text = raw.trim()
                    if (text.isEmpty()) continue
                    text.toLongOrNull()?.let { return normalizeEpoch(it) }
                    text.toDoubleOrNull()?.let { return normalizeEpoch(it.toLong()) }
                    parseDateText(text)?.let { return it }
                }
            }
        }
        return null
    }

    private fun normalizeEpoch(value: Long): Long =
        if (value in 1..MAX_SECONDS_EPOCH) value * 1000L else value

    private fun parseDateText(text: String): Long? {
        val patterns = listOf(
            "yyyy-MM-dd'T'HH:mm:ss.SSSXXX",
            "yyyy-MM-dd'T'HH:mm:ssXXX",
            "yyyy-MM-dd'T'HH:mm:ss",
            "yyyy-MM-dd HH:mm:ss",
            "yyyy-MM-dd",
        )
        for (pattern in patterns) {
            val parsed = runCatching {
                SimpleDateFormat(pattern, Locale.US).apply {
                    timeZone = TimeZone.getTimeZone("UTC")
                }.parse(text)?.time
            }.getOrNull()
            if (parsed != null) return parsed
        }
        return null
    }

    /**
     * Duracion en milisegundos.
     *
     * `totalTime` ya viene en milisegundos (la exportacion divide por 1000 para obtener segundos);
     * las claves alternativas (`duration`, `exerciseTime`) suelen venir en segundos, asi que se
     * distinguen por magnitud: menos de un dia se interpreta como segundos.
     */
    private fun readDuration(node: JSONObject): Long? {
        readDouble(node, "totalTime", "total_time")?.let { return it.toLong() }
        readDouble(node, "durationMs", "duration_ms")?.let { return it.toLong() }
        readDouble(node, "duration", "exerciseTime")?.let { raw ->
            return if (raw < SECONDS_PER_DAY) (raw * 1000).toLong() else raw.toLong()
        }
        return null
    }

    private fun readString(node: JSONObject, vararg keys: String): String? {
        for (key in keys) {
            val value = node.opt(key)
            if (value is String && value.isNotBlank()) return value.trim()
            if (value is Number) return value.toString()
        }
        return null
    }

    private fun readInt(node: JSONObject, vararg keys: String): Int? =
        readDouble(node, *keys)?.roundToInt()

    private fun readDouble(node: JSONObject, vararg keys: String): Double? {
        for (key in keys) {
            val value = node.opt(key)
            when (value) {
                is Number -> return value.toDouble()
                is String -> value.trim().replace(',', '.').toDoubleOrNull()?.let { return it }
            }
        }
        return null
    }

    /** Epoch en segundos del siglo XXI: por debajo de este valor se asume segundos, no milisegundos. */
    private const val MAX_SECONDS_EPOCH = 4_102_444_800L

    /** Segundos de un dia: separa una duracion en segundos de una en milisegundos. */
    private const val SECONDS_PER_DAY = 86_400.0

    /** Marca de origen en la nota de una sesion importada. */
    const val NOTE_PREFIX = ImportedWorkoutNotes.HUAWEI

    /** `true` si la sesion vino de una importacion (el origen queda en la nota). */
    fun isImportedNote(notes: String?): Boolean = ImportedWorkoutNotes.isImported(notes)

    /** Nota de la sesion importada: origen y los datos que Huawei si registro. */
    fun noteFor(workout: ImportedWorkout): String = ImportedWorkoutNotes.noteFor(workout)
}
